python/functional-programming/sum-all-map-functions/map-built-in-imperative.md

```diff
@@ -0,0 +1,38 @@
+---
+author: stefkn
+type: normal
+category: must-know
+---
+
+# The map Function
+
+
+---
+
+## Content
+
+Let's say we have a list of point scores in a game, and we want to add 2 to every score. In the imperative programming style, we would do something like this:
+
+```python
+scores = [1, 0, 4, 5]
+for i in range(0, len(scores)):
+  scores[i] = scores[i] + 2
+```
+
+But what if we want to apply this transformation to another list?
+
+How can we keep track of whether the transformation has been applied or not?
+
+These issues can make imperative programming[1] hard to work with.
+
+
+
+---
+
+## Footnotes
+
+
+[1:Imperative Programming]
+Imperative programming, as opposed to declarative programming (of which functional programming is a subtype), is the programming paradigm where statements are used to change a program's state. The imperative style requires certain mechanisms such as direct data modification (mutability) and looping statements. These mechanisms sometimes detrimentally affect readability and efficiency.
+
+An easy rule of thumb is that imperative programming specifies **how** (what steps are taken during execution) a program should operate. This is in contrast to declarative languages which focus on **what** is to be computed instead.
```

Comment:
Maybe take out the essential from these paragraphs?

```suggestion
Imperative code specifies each step towards computing a result.

Declarative code specifies the result but not each individual step.
```

Reply:
We could use a real-life analogy too:

#### Baking a Cake Analogy (Imperative Programming vs Declarative Programming)

Imperative Programming: It's like having a detailed recipe for a cake and you're instructing someone step by step on how to bake it. You tell them to preheat the oven to a certain temperature, to mix the ingredients in a specific order, to grease the pan, pour the batter in, bake for a specific time, and so on. If any step is left out or done incorrectly, the cake may not turn out as expected.

Declarative Programming: On the other hand, it's like going to a bakery and just ordering the cake. You tell them what you want (a chocolate cake, for instance) and you don't care about the steps they take to bake it. You're just declaring your desired end result and they figure out the steps to achieve it.

#### Road Trip Analogy (Imperative Programming vs Declarative Programming)

Imperative Programming: Imagine planning a road trip where you're the driver and you list down every action you need to take: start the car, turn right after 200 yards, go straight for 2 miles, turn left at the roundabout, continue for another mile, stop at the gas station, fill the tank, and so on. You control each and every action and if you miss a turn, you'll need to correct it yourself.

Declarative Programming: Now, compare this with using a GPS navigation system where you just input your destination. The system figures out the best route, where to turn, when to stop for gas, and so on. If you make a wrong turn, it recalculates and provides new instructions. You're not telling it every step to take, you're just declaring your end goal and the system figures out how to get there.

Reply:
A few paragraphs from which you could take inspiration:

Imperative Programming:

Imperative programming is a paradigm where you, the programmer, tell the computer how to do something. That is, you provide explicit instructions for each step that needs to be taken to achieve a certain goal. These instructions are typically executed in a sequence, and often include control structures such as loops, conditionals, and subroutines. This is a step-by-step approach, much like a recipe, where you provide a detailed list of what needs to be done.

Imperative programming is generally more low-level and detailed, and many common languages such as C, C++, and Python have imperative elements. However, this also means that imperative code can be more complex and harder to maintain, especially for larger programs.

Declarative Programming:

In contrast, declarative programming is a paradigm where you tell the computer what you want to do, and let the computer figure out how to do it. Instead of providing a step-by-step "recipe", you declare a desired result, and the computer takes care of the details of how to achieve that result.

Declarative programming often results in code that is more concise and easier to read, as it focuses on higher-level logic rather than implementation details. Examples of declarative languages include SQL (for database queries), HTML (for webpage structure), and Prolog (for logical programming).

In practice, many modern programming languages and frameworks use a combination of both imperative and declarative programming, allowing programmers to use the best tool for each job. For instance, a JavaScript developer might use imperative programming to manipulate variables and control flow, while also using declarative programming to manipulate the DOM or make queries to a database.